### PR TITLE
fix(query-builder): make sure `$or` and `$and` combined works correctly

### PR DESCRIPTION
**Pull request by charmpitz**
Fix for https://github.com/mikro-orm/mikro-orm/issues/792

### CODE REVIEW

**Comment by B4nan** [comment]
Thanks!

File: tests/QueryBuilder.test.ts [hunk 1]
```diff
@@ -1240,6 +1240,30 @@ describe('QueryBuilder', () => {
...
+    });
+    expect(qb3.getQuery()).toEqual(
+      'select `a`.* from `author2` as `a` ' +
+      'where (`a`.`name` = ?) and (`a`.`name` = ?)');
+
+
+    const qb4 = orm.em.createQueryBuilder(Author2, 'a');
+    qb4.select('*').where({
+      $and: [
+        { $or: [{ name: 'value1', email: 'value2' }] },
```
Comment (B4nan) [inline tests/QueryBuilder.test.ts:1259]: uh wait a minute, this looks wrong. you have single item in the `$or` array, which is object. the objet properties should be `$and`ed, not `$or`ed.

this should produce this query:

```sql
select `a`.* from `author2` as `a` 
where (`a`.`name` = ? and `a`.`email` = ?) and (`a`.`name` = ? and `a`.`email` = ?)
```

to build the query you are asserting, you would have to have multiple items in the `$or` array:

```typescript
    const qb4 = orm.em.createQueryBuilder(Author2, 'a');
    qb4.select('*').where({
      $and: [
        { $or: [{ name: 'value1' }, { email: 'value2' }] }, // now we have two objects in the $or array
        { $or: [{ name: 'value3' }, { email: 'value4' }] }, // now we have two objects in the $or array
      ],
    });
```

Comment (charmpitz) [reply tests/QueryBuilder.test.ts:1259]: Indeed, you're right, I'll fix it in a new PR and incorporate a few more tests to make sure it works accordingly. Let me know if you have more insights :v:

Comment (charmpitz) [reply tests/QueryBuilder.test.ts:1259]: I had hoped this would be easier to fix but unfortunately it took me all day to make it work properly. I've added more test case scenarios and I've found a lot more corner cases. Before moving further there are still a few quirks we should discuss:
1. There are extra lots of extra parenthesis and frankly I don't think we can get rid of them. They seem to be the main reason why the queries weren't working properly. Because of this, sadly, there are lots of tests that need to be fixed as well..
2. I'm not sure how `method` argument should be treated

I've added a PR (https://github.com/mikro-orm/mikro-orm/pull/798), let me know what you think. It would probably be best to play around with it too.

Comment (B4nan) [reply tests/QueryBuilder.test.ts:1259]: Thanks, will take a look at what you did in more detail and see if I can finalize that. Btw its better to move the conversation to the new PR.

Comment (charmpitz) [reply tests/QueryBuilder.test.ts:1259]: You're welcome, let me know if I can be of any help.
